urllib3/connectionpool.py

```diff
@@ -576,6 +577,19 @@ def urlopen(self, method, url, body=None, headers=None, retries=None,
         # ensures we do proper cleanup in finally.
         clean_exit = False
 
+        # In the event of a redirect or retry, we need to set a file-like
+        # body's read position back to the original point. If there is no
+        # value stored, let's try to get one for future use.
+        if body_pos is not None:
+            rewind_body(body, body_pos)
+        elif getattr(body, 'tell', None) is not None:
+            try:
+                body_pos = body.tell()
+            except (IOError, OSError):
+                # This differentiates from None, allowing us to catch
+                # a failed `tell()` later when trying to rewind the body.
+                body_pos = object()
```

Comment:
So just to clarify, you're talking about moving lines 589-591 into utils/requests.py, yes? I'm not sure we want to move much more than that out of `urlopen`.